pkg/metrics/metrics.go

```diff
@@ -96,19 +97,32 @@ func WithInstallIDFileOverride(path string) Option {
 	}
 }
 
+// WithAsyncErrorHandler registers a function to handle any errors occurring
+// during async metric writing. It may be run after the context has
+// timed out, so long-lived blocking operations are discouraged.
+func WithAsyncErrorHandler(handler func(ctx context.Context, err error)) Option {
+	return func(o *options) *options {
+		o.errorHandler = handler
+		return o
+	}
+}
+
 // MetricWriter is a client for reporting metrics about an application's usage.
 type MetricWriter interface {
 	WriteMetric(ctx context.Context, name string, count int64) error
-	WriteMetricAsync(ctx context.Context, name string, count int64) func() error
+	WriteMetricAsync(ctx context.Context, name string, count int64)
+	Close()
 }
 
 type client struct {
-	AppID      string
-	AppVersion string
-	InstallID  string
-	HTTPClient *http.Client
-	OptOut     bool
-	Config     *metricsConfig
+	AppID        string
```

Comment:
I see, it's a pattern I've been using on unexported structs since you can't serialize unless the fields are exported.

In this case we aren't serializing, so I guess it doesn't make sense.

Reply:
Changed in a separate pr.